### PR TITLE
oauth: encrypted storage

### PR DESCRIPTION
**Pull request by spolu**
## Description

Implements the functions that will let us store encrypted secrets for OAuth connections

## Risk

N/A not deployed  yet

## Deploy Plan

N/A

### CODE REVIEW

File: core/src/oauth/connection.rs [hunk 1]
```diff
@@ -182,6 +202,96 @@ impl Connection {
         &self.metadata
     }
 
+    pub fn access_token_expiry(&self) -> Option<u64> {
+        self.access_token_expiry
+    }
+
+    pub fn access_token_expiry_or_default(&self) -> Option<u64> {
+        self.access_token_expiry
+    }
```
Comment (flvndvd) [inline core/src/oauth/connection.rs:211]: Not sure to get the behavior difference from those two methods 🤔

File: core/bin/oauth_generate_key.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+use base64::{engine::general_purpose, Engine as _};
```
Comment (flvndvd) [inline core/bin/oauth_generate_key.rs:1]: So the idea, it to use this script once to generate the encryption key before setting it as an env variable, is that right?

File: core/src/oauth/connection.rs [hunk 3]
```diff
@@ -3,9 +3,29 @@ use crate::utils;
...
+};
 use serde::{Deserialize, Serialize};
-use std::fmt;
 use std::str::FromStr;
+use std::{env, fmt};
+
+lazy_static! {
+    static ref ENCRYPTION_KEY: Result<aead::LessSafeKey, &'static str> = {
+        let encoded_key =
+            env::var("OAUTH_ENCRYPTION_KEY").map_err(|_| "OAUTH_ENCRYPTION_KEY: not set")?;
```
Comment (flvndvd) [inline core/src/oauth/connection.rs:20]: So the encryption key is set in the environment as a base64?

**Review by flvndvd — COMMENTED** [review]
LGTM, before stamping just want to make sure I get the full picture.

File: core/src/oauth/connection.rs (hunk 1 above)
Comment (spolu) [reply core/src/oauth/connection.rs:211]: noise to be removed

File: core/bin/oauth_generate_key.rs (hunk 2 above)
Comment (spolu) [reply core/bin/oauth_generate_key.rs:1]: Yeah can be removed once set in dev and prod

File: core/src/oauth/connection.rs (hunk 3 above)
Comment (spolu) [reply core/src/oauth/connection.rs:20]: yes as per the bin to generate it

**Comment by spolu** [comment]
Moving forward with it as this is all dev only for now

**Comment by spolu** [comment]
Happy to tweak things further based on review :+1:

File: core/src/oauth/connection.rs (hunk 3 above)
Comment (flvndvd) [reply core/src/oauth/connection.rs:20]: Yep that's what I figured but given that it was not explicit in the PR description or the spec, just wanted to confirm 🙃.

File: core/src/oauth/connection.rs [hunk 4]
```diff
@@ -113,23 +133,23 @@ impl Connection {
...
             created,
             provider,
             status,
             metadata,
-            authorization_code,
-            access_token,
             access_token_expiry,
-            refresh_token,
-            raw_json,
+            encrypted_authorization_code,
```
Comment (flvndvd) [inline core/src/oauth/connection.rs:149]: Just wondering if we need to encrypt all of those fields? I mean `access_token` and `refresh_token` obviously but the other 🤔.

Do you have an example of what `raw_json` contains from your test?

Comment (spolu) [reply core/src/oauth/connection.rs:149]: Can be seen in nango.dev. It contains the access token and refresh token. We could store a sanitized version but that introduces more logic that can break with time. So instead the approach is to store it encrypted and expose a whiltelist of fields on the Connection (user related in general)

```
{
    "ok": true,
    "app_id": "A055TCFUZND",
    "authed_user": {
        "id": "U04VCD0FRKR"
    },
    "scope": "app_mentions:read,channels:history,channels:join,channels:read,chat:write,groups:read,im:history,im:read,mpim:read,team:read,users:read,users:read.email,metadata.message:read",
    "token_type": "bot",
    "access_token": "xoxb-...",
    "bot_user_id": "U055TFM6BU1",
    "team": {
        "id": "T050RH73H9P",
        "name": "Dust"
    },
    "enterprise": null,
    "is_enterprise_install": false
}
```

**Review by flvndvd — COMMENTED** [review]
LGTM ✅

**Pull request merged**